EasyCommands.Tests/ScriptTests/ScriptTest.cs

```diff
@@ -0,0 +1,182 @@
+        /// <param name="groupName">The name of the group being mocked.</param>
+        /// <param name="blockMocks">The blocks being mocked and that will be returned later.</param>
+        public void MockBlocksInGroup<T>(String groupName, params Mock<T>[] blockMocks) where T : class, IMyTerminalBlock
+        {
+            var mockGroup = new Mock<IMyBlockGroup>();
+            mockGroup.Setup(g => g.Name).Returns(groupName);
+            mockGroup.Setup(g => g.GetBlocksOfType(It.IsAny<List<T>>(), It.IsAny<Func<T, bool>>()))
+                .Callback((Action<List<T>, Func<T, bool>>)GetBlocksOfTypeCallback);
+            mockedGroups.Add(mockGroup.Object);
+
+            SetupMockBlocksByType(blockMocks);
+        }
+
+        private void GetBlocksOfTypeCallback<T>(List<T> blocks, Func<T, bool> collect) where T : class, IMyTerminalBlock
+        {
+            // Find the appropriate blocks by type and then invoke the given collector function on them as a filter
+            blocks.AddRange(mockedBlocksByName.GetValue<T>(typeof(T))
+                .Select(x => x)
+                .Where(collect != null ? collect : (x) => true)
+                .ToList());
+        }
+
+        private void GetBlockGroupsCallback(List<IMyBlockGroup> groups, Func<IMyBlockGroup, bool> collect)
+        {
+            // Invoke the given collector function on all groups and return those groups matching
+            groups.AddRange(mockedGroups
+                .Select(x => x)
+                .Where(collect != null ? collect : (x) => true)
+                .ToList());
+        }
+
+        public void Dispose()
+        {
+            // TODO: Unset any static changes, etc
+            Program.STATE = ProgramState.STOPPED;
+            Program.LOG_LEVEL = LogLevel.INFO;
+        }
+
+        private void SetupMockBlocksByType<T>(string name, Mock<T>[] blockMocks) where T : class, IMyTerminalBlock
+        {
+            // Setup the mock blocks to mock their custom name to match the name provided
+            List<T> blockObjects = blockMocks.Select(b =>
+            {
+                b.Setup(x => x.CustomName).Returns(name);
+                return b.Object;
+            }).ToList();
+            // Store the mock block objects by their type
+            mockedBlocksByName.Add(typeof(T), blockObjects);
+        }
+
+        private void SetupMockBlocksByType<T>(Mock<T>[] blockMocks) where T: class, IMyTerminalBlock
+        {
+            // Store the mock block objects by their type
+            mockedBlocksByName.Add(typeof(T), blockMocks.Select(b => b.Object).ToList());
+        }
+
+        public class GenericListDictionary
+        {
+            private Dictionary<Type, List<Object>> dictionary = new Dictionary<Type, List<object>>();
+
+            public void Add<T>(Type type, List<T> values)
+            {
+                dictionary.Add(type, values.Select(x => (Object)x).ToList());
```

Comment:
I think this will fail if you try to add to the dictionary twice.

Instead, I think you need to fetch the current list by key (adding an empty list if not exists) and then appending the objects to the given list.  That way we can mock multiple blocks of the same type with different custom names and/or belonging to different groups.

Reply:
That's right! I wanted to write a test that exercised a variety of scenarios like multiple blocks of the same type. Ran out of steam the other night but I've added that now (and addressed this issue).